### PR TITLE
HID Mouse with absolute positioning

### PR DESCRIPTION
**Pull request by tobozo**
The current mouse support is limited to relative positioning.

This PR adds absolute mouse descriptor template and descriptors.

### CODE REVIEW

File: src/class/hid/hid.h
Comment (me-no-dev) [inline src/class/hid/hid.h:313]: to follow the naming of the rest of the structs, this should be renamed to `hid_abs_mouse_report_t`

**Comment by FallingSnow** [comment]
@tobozo Please fix test errors.

**Comment by tobozo** [comment]
@FallingSnow fetching upstream fixed it on my fork

https://github.com/tobozo/tinyusb/runs/6149927028?check_suite_focus=true

**Review by tao-j — APPROVED** [review]
LGTM

**Review by earlephilhower — CHANGES_REQUESTED** [review]
@tobozo this is great, but if `abs_mouse` is to be a first-class citizen in TinyUSB, shouldn't we add a couple helper functions to hid_device.h to match the base mouse?

Like:
````
bool tud_hid_n_abs_mouse_report(uint8_t instance, uint8_t report_id, uint8_t buttons, int16_t x, int16_t y, int8_t vertical, int8_t horizontal);`

static inline bool tud_hid_abs_mouse_report(uint8_t report_id, uint8_t buttons, int16_t x, int16_t y, int8_t vertical, int8_t horizontal)
{
  return tud_hid_n_abs_mouse_report(0, report_id, buttons, x, y, vertical, horizontal);
}
````

**Comment by tobozo** [comment]
@earlephilhower thanks for the suggestion :+1:  should I also add the C function in hid_device.c ?

```C
bool tud_hid_n_abs_mouse_report(uint8_t instance, uint8_t report_id, uint8_t buttons, int16_t x, int16_t y, int8_t vertical, int8_t horizontal)
{
  hid_abs_mouse_report_t report =
  {
    .buttons = buttons,
    .x       = x,
    .y       = y,
    .wheel   = vertical,
    .pan     = horizontal
  };
  return tud_hid_n_report(instance, report_id, &report, sizeof(report));
}
```

**Comment by earlephilhower** [comment]
Yes, that would be needed since we only have the prototype otherwise which will make the linker unhappy. :)

**Comment by tobozo** [comment]
ideally the touch zone should have a shape proportional to the target screen so it's not exactly as intuitive as using the relative mouse, especially when the desktop has multiple screens :sweat:  

there are a lot of use cases anyway, I remember exploring those with an ESP32-S2:

- physical mouse
- physical touchpad (e.g. using a XPT2046)
- mouse server (e.g. forwards mouse reports from a http client to the host)
- scripted mouse (e.g. [drawing hand in an image editor](https://www.youtube.com/watch?v=Nsf1brl3z0w))

**Comment by doge-gif** [comment]
@earlephilhower May I know what's blocking this merge?

**Comment by earlephilhower** [comment]
@doge-gif this isn't my repo or PR, so you'll need to ask @hathach . 😄    I've implemented it in the [arduino-pico](https://github.com/earlephilhower/arduino-pico) core already and will undo my private hacks once this PR merges and the Raspberry Pi team updates their SDK to include it.

**Comment by tobozo** [comment]
@me-no-dev does this need further review? Looks like it's missing your approval

poke @hathach 

image

**Review by me-no-dev — APPROVED** [review]

**Comment by me-no-dev** [comment]
Sorry @tobozo ! Done.

**Comment by ithinkido** [comment]
Please can this get merged

**Comment by thewh1teagle** [comment]
What's the status of this PR? is it going to be merged?

**Comment by hathach** [comment]
Unfortunately, I still havent got time to look at this. Will do whenever possible

**Comment by thewh1teagle** [comment]
It's nearly 50 lines and a year ago :)
Hope you get time to check it out.
By the way, can we use the same device for absolute and relative in the same descriptor or it has to be 2 different devices?

**Review by hathach — APPROVED** [review]
thank you very much for the PR and sorry for late response. I kind of forgot about these in my pile of email.

**Pull request merged**
